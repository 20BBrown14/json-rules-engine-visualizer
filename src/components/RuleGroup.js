import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Button } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { SQFormDropdown, useSQFormContext } from '@selectquotelabs/sqform';
import { EXPRESSION_SQFORMDROPDOWN_OPTIONS, GROUP_TYPE, RULE_TYPE } from '../constants/constants';

const groupGridStyles = makeStyles({
  groupGrid: {
    background: 'rgb(233, 241, 241)',
    paddingBottom: '20px',
    border: '1px solid #b3b3b3',
    width: '100%',
    marginBottom: '25px',
  },
  nestedGroupGrid: {
    width: '100%',
    margin: 0,
  },
  groupButton: {
    margin: '0 3px',
  },
});

function RuleGroup({
  ruleGroupName,
  addChildToGroup,
  isTopGroup = false,
  children,
}) {
  const groupClasses = groupGridStyles();

  const { initialValues } = useSQFormContext();

  // Check if the initial values for this group exist yet
  const initialValue = Object.keys(initialValues).find((key) => (
    `${ruleGroupName}_expression` === key
  ));

  // If initial value doesn't exist don't render anything
  if (!initialValue) {
    return null;
  }

  return (
    <Grid container spacing={2} justify="flex-start" className={groupClasses.groupGrid}>
      <SQFormDropdown
        id={`${ruleGroupName}_dropdown`}
        name={`${ruleGroupName}_expression`}
        label="Boolean Expression"
        size={3}
      >
        {EXPRESSION_SQFORMDROPDOWN_OPTIONS}
      </SQFormDropdown>
      <Grid item sm={6}>
        <Button
          id={`${ruleGroupName}_addGroupButton`}
          className={groupClasses.groupButton}
          variant="contained"
          size="small"
          onClick={() => { addChildToGroup(ruleGroupName, GROUP_TYPE); }}
        >
          Add Group
        </Button>
        <Button
          className={groupClasses.groupButton}
          variant="contained"
          size="small"
          onClick={() => { addChildToGroup(ruleGroupName, RULE_TYPE); }}
        >
          Add Rule
        </Button>
        {!isTopGroup && (
        <Button
          className={groupClasses.groupButton}
          variant="contained"
          size="small"
        >
          Remove
        </Button>
        )}
      </Grid>
      <Grid item sm={3} />
      <Grid item sm={1} />
      <Grid item sm={11}>
        <Grid container spacing={4} className={groupClasses.nestedGroupGrid}>
          {children}
        </Grid>
      </Grid>
    </Grid>
  );
}

RuleGroup.propTypes = {
  /** Name of the rule group */
  ruleGroupName: PropTypes.string.isRequired,
  /** Add child to group function */
  addChildToGroup: PropTypes.func.isRequired,
  /** Whether this group is the top group */
  isTopGroup: PropTypes.bool,
  /** The rules or group that are a part of this group */
  children: PropTypes.node,
};

export default RuleGroup;
