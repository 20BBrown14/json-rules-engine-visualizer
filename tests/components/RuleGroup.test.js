import React from 'react';
import RuleGroup from '../../src/components/RuleGroup';
import SQFormWrapper from '../util/SQFormWrapper';

describe('RuleGroup', () => {
  it('renders properly', () => {
    const testRuleGroup = render(
      <SQFormWrapper initialValues={{ someName_expression: '' }}>
        <RuleGroup ruleGroupName="someName" />
      </SQFormWrapper>,
    );
    expect(testRuleGroup).toMatchSnapshot();
  });
});
