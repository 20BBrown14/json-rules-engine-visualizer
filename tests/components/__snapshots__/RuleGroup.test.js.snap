// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleGroup renders properly 1`] = `
<form
  action="#"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      class="MuiGrid-root makeStyles-groupGrid-1 MuiGrid-container MuiGrid-spacing-xs-2"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-3"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
          id="boolean expression"
        >
          Boolean Expression
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="boolean expression mui-component-select-someName_expression"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
            id="mui-component-select-someName_expression"
            role="button"
            tabindex="0"
          >
            - -
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            name="someName_expression"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
        <p
          class="MuiFormHelperText-root"
        >
          <span>
            â€‹
          </span>
        </p>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-groupButton-3 MuiButton-containedSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add Group
          </span>
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-groupButton-3 MuiButton-containedSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add Rule
          </span>
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-groupButton-3 MuiButton-containedSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Remove
          </span>
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-3"
      />
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-1"
      />
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-11"
      >
        <div
          class="MuiGrid-root makeStyles-nestedGroupGrid-2 MuiGrid-container MuiGrid-spacing-xs-4"
        />
      </div>
    </div>
  </div>
</form>
`;
